diff --git a/kernel/sched/core.c b/kernel/sched/core.c
index 71c819c62..c915cce55 100644
--- a/kernel/sched/core.c
+++ b/kernel/sched/core.c
@@ -12,6 +12,9 @@
 
 #include "sched.h"
 
+/* Added import for project 3*/
+#include <linux/syscalls.h>
+
 #include <linux/nospec.h>
 
 #include <linux/kcov.h>
@@ -7091,6 +7094,38 @@ SYSCALL_DEFINE1(nice, int, increment)
 
 #endif
 
+/* 
+ * sys_set_inactive - Have the thread mark itself as active or inactive
+ * @is_inactive: Flag to set status as active (0) or inactive (1)
+ * Inactive tasks are deprioritized by the CFS. put_prev_entity inflates vruntime on every preemption,
+ * check_preempt_tick forces the task of CPU at tick boundaries, check_preempt_wakeup yields to active waking tasks,
+ * pick_next_entity prefers active tasks when both exist on the runqueue.
+ */
+SYSCALL_DEFINE1(set_inactive, int, is_inactive)
+{
+	struct task_struct *task = current;
+	
+	// If the flag is not one of the recognized values, return error
+	if(is_inactive != 0 && is_inactive != 1)
+		return -EINVAL; 
+
+	// Case for reactivating a task. 
+	if (is_inactive == 0 && task->is_inactive == 1) {
+		struct rq_flags rf;
+		struct rq *rq = task_rq_lock(task, &rf);
+		task->se.vruntime = rq->cfs.min_vruntime; // Put task back at the front and clear
+											//  vruntime punishments occurred from being inactive.
+		task->is_inactive = 0;
+		task_rq_unlock(rq, task, &rf);
+	} else {
+		task->is_inactive = is_inactive;
+		if (is_inactive)
+			cond_resched(); // Switch if there is a higher priority task waiting, otherwise keep running
+	}
+
+	return 0;
+}
+
 /**
  * task_prio - return the priority value of a given task.
  * @p: the task in question.
